File: src/mrb_narya.c
/*
** mrb_narya.c - Narya class
**
** Copyright (c) KAGEYAMA Katsuhiko 2019
**
** See Copyright Notice in LICENSE
*/

#include "mruby.h"
#include "mruby/data.h"
#include "mruby/class.h"
#include "mruby/value.h"
#include "mrb_narya.h"

#define DONE mrb_gc_arena_restore(mrb, 0);

void mrb_mruby_esp32_narya_gem_init(mrb_state *mrb)
{
  mrb_narya_init_resouce(mrb);

  struct RClass *narya_module = mrb_define_module(mrb, "Narya");

  // Display
  struct RClass *display_module = mrb_define_module_under(mrb, narya_module, "Display");
  mrb_define_module_function(mrb, display_module, "width", mrb_narya_display_width, MRB_ARGS_NONE());
  mrb_define_module_function(mrb, display_module, "height", mrb_narya_display_height, MRB_ARGS_NONE());

  mrb_define_module_function(mrb, display_module, "draw_circle", mrb_narya_display_draw_circle, MRB_ARGS_REQ(4));
  mrb_define_module_function(mrb, display_module, "draw_line", mrb_narya_display_draw_line, MRB_ARGS_REQ(5));
  mrb_define_module_function(mrb, display_module, "draw_rect", mrb_narya_display_draw_rect, MRB_ARGS_ARG(5,1));
  mrb_define_module_function(mrb, display_module, "draw_pixel", mrb_narya_display_draw_pixel, MRB_ARGS_REQ(3));

  mrb_define_module_function(mrb, display_module, "clear", mrb_narya_display_clear, MRB_ARGS_NONE());
  mrb_define_module_function(mrb, display_module, "double_buffered?", mrb_narya_display_double_buffered, MRB_ARGS_NONE());
  mrb_define_module_function(mrb, display_module, "swap", mrb_narya_display_swap, MRB_ARGS_NONE());

  mrb_define_module_function(mrb, display_module, "draw_picture", mrb_narya_display_draw_picture, MRB_ARGS_REQ(3));
  mrb_define_module_function(mrb, display_module, "load_bitmap", mrb_narya_display_load_bitmap, MRB_ARGS_NONE());
  mrb_define_module_function(mrb, display_module, "scroll", mrb_narya_display_scroll, MRB_ARGS_REQ(2));

  mrb_define_module_function(mrb, display_module, "font_size", mrb_narya_display_font_size, MRB_ARGS_REQ(2));
  mrb_define_module_function(mrb, display_module, "draw_text", mrb_narya_display_draw_text, MRB_ARGS_ARG(4,1));

  // Input
  struct RClass *input_module = mrb_define_module_under(mrb, narya_module, "Input");
  mrb_define_module_function(mrb, input_module, "available", mrb_narya_input_available, MRB_ARGS_NONE());
  mrb_define_module_function(mrb, input_module, "keydown?", mrb_narya_input_keydown, MRB_ARGS_REQ(1));
  mrb_define_module_function(mrb, input_module, "get_key", mrb_narya_input_get_key, MRB_ARGS_NONE());
  mrb_define_module_function(mrb, input_module, "paddown?", mrb_narya_input_pad_down, MRB_ARGS_REQ(1));

  //struct RClass *image_module = mrb_define_module_under(mrb, narya_module, "Image");

  // Sprite
  struct RClass *sprite_class = mrb_define_class_under(mrb, narya_module, "Sprite", mrb->object_class);
  MRB_SET_INSTANCE_TT(sprite_class, MRB_TT_DATA);
  mrb_define_method(mrb, sprite_class, "initialize", mrb_narya_sprite_initialize, MRB_ARGS_REQ(1));
  mrb_define_method(mrb, sprite_class, "move_to", mrb_narya_sprite_move_to, MRB_ARGS_REQ(2));
  mrb_define_method(mrb, sprite_class, "move", mrb_narya_sprite_move, MRB_ARGS_REQ(2));
  
  // Bitmap
  struct RClass *bitmap_class = mrb_define_class_under(mrb, narya_module, "Bitmap", mrb->object_class);
  MRB_SET_INSTANCE_TT(bitmap_class, MRB_TT_DATA);
  mrb_define_method(mrb, bitmap_class, "initialize", mrb_narya_bitmap_initialize, MRB_ARGS_NONE());
  mrb_define_method(mrb, bitmap_class, "load", mrb_narya_bitmap_load, MRB_ARGS_REQ(1));
  mrb_define_method(mrb, bitmap_class, "draw", mrb_narya_bitmap_draw, MRB_ARGS_REQ(2));

  // Config
  struct RClass *config_module = mrb_define_module_under(mrb, narya_module, "Config");
  mrb_define_module_function(mrb, config_module, "firmware_ver", mrb_narya_config_firmware_ver, MRB_ARGS_NONE());
  mrb_define_module_function(mrb, config_module, "narya_ver", mrb_narya_config_narya_ver, MRB_ARGS_NONE());

  // Sound
  struct RClass *sound_module = mrb_define_module_under(mrb, narya_module, "Sound");
  mrb_define_module_function(mrb, sound_module, "load_mml", mrb_narya_sound_load, MRB_ARGS_REQ(1));
  mrb_define_module_function(mrb, sound_module, "play", mrb_narya_sound_play, MRB_ARGS_NONE());
  mrb_define_module_function(mrb, sound_module, "stop", mrb_narya_sound_stop, MRB_ARGS_NONE());



  /** To be implemented...
  * Display module
  *  Draw Ellipse
  *  Copy Rect
  *  Draw Draw Path(fill)
  *  double_buffer?
  * Storage module
  *  open
  *  read
  *  write
  *  close
  * Sprite class
  *  add_image
  *  set_config
  **/

  DONE;
}

void mrb_mruby_esp32_narya_gem_final(mrb_state *mrb)
{
  mrb_narya_finalize_resouce(mrb);
}

File: mrblib/mrb_narya.rb
module Narya
  class Color
    attr_accessor :r,:g,:b
    def initialize(r,g,b)
      @r=r
      @g=g
      @b=b
    end
    def to_s
      "#{r},#{g},#{b}"
    end
  end

  module Display
  end

  module Input
  end

  module Image
  end

  module Sound
  end

  class Bitmap
    attr_reader :width,:height
  end

  class Sprite
  end

  module Key
    K_NONE   = 0
    K_SPACE  = 1
    K_0      = 2
    K_1      = 3
    K_2      = 4
    K_3      = 5
    K_4      = 6
    K_5      = 7
    K_6      = 8
    K_7      = 9
    K_8      = 10
    K_9      = 11
    K_a      = 22
    K_b      = 23
    K_c      = 24
    K_d      = 25
    K_e      = 26
    K_f      = 27
    K_g      = 28
    K_h      = 29
    K_i      = 30
    K_j      = 31
    K_k      = 32
    K_l      = 33
    K_m      = 34
    K_n      = 35
    K_o      = 36
    K_p      = 37
    K_q      = 38
    K_r      = 39
    K_s      = 40
    K_t      = 41
    K_u      = 42
    K_v      = 43
    K_w      = 44
    K_x      = 45
    K_y      = 46
    K_z      = 47
    K_A      = 48
    K_B      = 49
    K_C      = 50
    K_D      = 51
    K_E      = 52
    K_F      = 53
    K_G      = 54
    K_H      = 55
    K_I      = 56
    K_J      = 57
    K_K      = 58
    K_L      = 59
    K_M      = 60
    K_N      = 61
    K_O      = 62
    K_P      = 63
    K_Q      = 64
    K_R      = 65
    K_S      = 66
    K_T      = 67
    K_U      = 68
    K_V      = 69
    K_W      = 70
    K_X      = 71
    K_Y      = 72
    K_Z      = 73
    K_ESCAPE         = 125
    K_RETURN         = 144
    K_UP             = 151
    K_DOWN           = 153
    K_LEFT           = 155
    K_RIGHT          = 157
  end
end
